gui/src/components/widgets/SkeletonVisualizerWidget.tsx

```diff
@@ -1,122 +1,132 @@
-  const leftHand = new Bone();
-  leftLowerArm.add(leftHand);
+export function OrthographicCameraWrapper() {
+  const { size } = useThree();
+  const aspect = useMemo(() => size.width / size.height, [size]);
 
-  const rightShoulder = new Bone();
-  neck.add(rightShoulder);
-  const rightUpperArm = new Bone();
-  rightShoulder.add(rightUpperArm);
-  const rightLowerArm = new Bone();
-  rightUpperArm.add(rightLowerArm);
-  const rightHand = new Bone();
-  rightLowerArm.add(rightHand);
+  return (
+    <OrthographicCamera
+      makeDefault
+      zoom={200}
+      top={frustumSize / factor}
+      bottom={frustumSize / -factor}
+      left={(0.5 * frustumSize * aspect) / -factor}
+      right={(0.5 * frustumSize * aspect) / factor}
+      near={0.1}
+      far={1000}
+      position={[25, 75, 50]}
+    />
+  );
+}
 
-  const leftUpperLeg = new Bone();
-  hip.add(leftUpperLeg);
-  const leftLowerLeg = new Bone();
-  leftUpperLeg.add(leftLowerLeg);
-  const leftFoot = new Bone();
-  leftLowerLeg.add(leftFoot);
+const yParts = [
+  BodyPart.NECK,
+  BodyPart.UPPER_CHEST,
+  BodyPart.CHEST,
+  BodyPart.WAIST,
+  BodyPart.HIP,
+  BodyPart.LEFT_UPPER_LEG,
+  BodyPart.LEFT_LOWER_LEG,
+];
 
-  const rightUpperLeg = new Bone();
-  hip.add(rightUpperLeg);
-  const rightLowerLeg = new Bone();
-  rightUpperLeg.add(rightLowerLeg);
-  const rightFoot = new Bone();
-  rightLowerLeg.add(rightFoot);
+export function SkeletonVisualizerWidget() {
+  const { bones: _bones } = useAppContext();
 
-  bones.push(
-    head,
-    neck,
-    chest,
-    waist,
-    hip,
-    leftUpperLeg,
-    rightUpperLeg,
-    leftLowerLeg,
-    rightLowerLeg,
-    leftFoot,
-    rightFoot,
-    leftLowerArm,
-    rightLowerArm,
-    leftUpperArm,
-    rightUpperArm,
-    leftHand,
-    rightHand,
-    leftShoulder,
-    rightShoulder
+  const bones = useMemo(
+    () => new Map(_bones.map((b) => [b.bodyPart, b])),
+    [JSON.stringify(_bones)]
   );
 
-  return new Skeleton(bones);
-})();
+  const skeleton = useRef<Bone[]>();
 
-export function SkeletonVisualizerWidget() {
-  const { bones } = useAppContext();
+  useEffect(() => {
+    skeleton.current = createChildren(bones, BoneKind.root);
+  }, [bones.size]);
 
-  const skeleton = useMemo(() => {
-    const skeleton = threeSkeleton.clone();
-    for (const solarBone of bones) {
-      const bone = skeleton.bones[solarBone.bodyPart - 1];
-      const localPos = bone.worldToLocal(
-        Vector3FromVec3fT(solarBone.headPositionG)
-      );
-      bone.position.set(localPos.x, localPos.y, localPos.z);
+  useEffect(() => {
+    skeleton.current?.forEach(
+      (bone) => bone instanceof BoneKind && bone.updateData(bones)
+    );
+  }, [bones]);
 
-      if (!bone.parent) {
-        bone.applyQuaternion(QuaternionFromQuatT(solarBone.rotationG));
-        continue;
-      }
+  const heightOffset = useMemo(() => {
+    const hmd = bones.get(BodyPart.HEAD);
+    if (hmd?.headPositionG?.y === undefined || hmd.headPositionG?.y > 0) {
```

Comment:
This looks really weird, why is it returning 0 if it's undefined? What's the height offset for, when the head position is unavailable?

Reply:
yep